### ISSUE TITLE
Websocket In "Connect to" fails with 404 when not passing origin header

### ISSUE THREAD

**Issue by svloosbroek** [node-red/node-red#1369]
Some websocket servers require the origin header to be sent, or else a 404 is returned and the node remains "Disconnected". The origin header is normally set by the webbrowser. This behaviour can be validated by altering line 22 in 22-websocket.js as follows: 

`var socket = new ws(node.path, {origin: "localhost"});`

Before passing the origin option to ws the connection would return a 404. Making this modification allows the node to connect successfully. Can be verified with the application otmonitor, available at http://otgw.tclcode.com/download.html

Suggestion: Please provide an option to pass the origin header when using the Connect to option

**Comment by knolleary** [comment]
If you wanted to open a pull request with this in we can review there.

**Comment by dceejay** [comment]
Would bad things happen if we always set that option (even if unnecessary)?

**Comment by svloosbroek** [comment]
It is technically an optional field, the [websocket RFC](https://tools.ietf.org/html/rfc6455) mentions:

> The |Origin| header field [RFC6454] is used to protect against
> unauthorized cross-origin use of a WebSocket server by scripts using
> the WebSocket API in a web browser.  The server is informed of the
> script origin generating the WebSocket connection request.  If the
> server does not wish to accept connections from this origin, it can
> choose to reject the connection by sending an appropriate HTTP error
> code.  This header field is sent by browser clients; for non-browser
> clients, this header field may be sent if it makes sense in the
> context of those clients.

Given the above I don't see why any websocket server would object to receiving this field in the handshake.

**Comment by svloosbroek** [comment]
Further considerations from the RFC below, the current implementation is correct as NR is not a browser client. For compatibility reasons one could add the option to pass the origin header but based on the RFC is it not required.

> The request MUST include a header field with the name |Origin|
> [RFC6454] if the request is coming from a browser client.  If
> the connection is from a non-browser client, the request MAY
> include this header field if the semantics of that client match
> the use-case described here for browser clients.  The value of
> this header field is the ASCII serialization of origin of the
> context in which the code establishing the connection is
> running.  See [RFC6454] for the details of how this header field
> value is constructed.
> 
> As an example, if code downloaded from www.example.com attempts
> to establish a connection to ww2.example.com, the value of the
> header field would be "http://www.example.com".

**Comment by hardillb** [comment]
True, but I can see a server filtering on an expected value, so just setting it to localhost may not be the best plan. I would make it optional and a configurable option adding it

**Comment by knolleary** [comment]
Closing as part of a tidy-up of the issue tracker — this has been inactive for a long time and much has changed in Node-RED since. If it's still relevant on a current version, please open a fresh issue with up-to-date details and we'll take a look. Thanks for raising it.
